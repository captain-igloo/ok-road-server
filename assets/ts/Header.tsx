import * as React from 'react';

import UserMenu from './UserMenu';

interface Props {
    showDemo?: boolean;
    showMap?: boolean;
    showRegister?: boolean;
    showSignIn?: boolean;
    user?: {
        fullName: string;
    };
}

export default function Header(props: Props) {
    const {
        showDemo,
        showMap,
        showRegister,
        showSignIn,
        user,
    } = props;

    return (
        <div className="header">
            <div className="logo">
                <a href="/">
                    <img alt="Home" src="/img/ok.svg" width="40px" height="40px" />
                </a>
                &nbsp;
                OK Road New Zealand
            </div>
            <UserMenu
                showDemo={showDemo}
                showMap={showMap}
                showRegister={showRegister}
                showSignIn={showSignIn}
                user={user}
            />
        </div>
    );
}
