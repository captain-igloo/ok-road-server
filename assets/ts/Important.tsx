import * as React from 'react';

interface Props {
    children: React.ReactNode;
}

export default function Important(props: Props) {
    const { children } = props;

    return (
        <div className="alert__important">
            <p className="alert__important--title">
                <img alt="" src="/img/important.svg" width={16} />
                {' '}
                Important
            </p>
            <p className="alert__important--body">
                {children}
            </p>
        </div>
    );
}
