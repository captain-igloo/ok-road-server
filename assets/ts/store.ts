import { combineReducers, configureStore } from '@reduxjs/toolkit';

import configReducer from './config/slice';
import friendsReducer from './friends/slice';
import mapReducer from './map/slice';
import notificationsReducer from './notifications/slice';

const rootReducer = combineReducers({
    config: configReducer,
    friends: friendsReducer,
    notifications: notificationsReducer,
    map: mapReducer,
});

export const setupStore = (preloadedState?: Partial<RootState>) => configureStore({
    reducer: rootReducer,
    preloadedState,
});

export type AppStore = ReturnType<typeof setupStore>;

export type RootState = ReturnType<typeof rootReducer>;

export type AppDispatch = AppStore['dispatch'];
