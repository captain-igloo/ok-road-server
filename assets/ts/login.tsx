import * as React from 'react';
import Alert from 'react-bootstrap/Alert';
import Button from 'react-bootstrap/Button';
import Card from 'react-bootstrap/Card';
import Form from 'react-bootstrap/Form';
import * as ReactDOM from 'react-dom/client';

import Header from './Header';

declare global {
    interface Window {
        error: string;
        token: string;
    }
}

const element = document.getElementById('login');

if (element) {
    const root = ReactDOM.createRoot(element);

    let alert;
    if (window.error) {
        alert = <Alert variant="danger">{window.error}</Alert>;
    }

    root.render(
        <>
            <Header showSignIn={false} />
            <Card className="login-container">
                <Card.Header>Sign In</Card.Header>
                <Card.Body>
                    {alert}
                    <Form method="post">
                        <Form.Group className="mb-3">
                            <Form.Label>Username</Form.Label>
                            <Form.Control name="_username" required type="text" />
                        </Form.Group>
                        <Form.Group className="mb-3">
                            <Form.Label>Password</Form.Label>
                            <Form.Control name="_password" required type="password" />
                        </Form.Group>
                        <input name="_csrf_token" type="hidden" value={window.token} />
                        <Button variant="primary" type="submit">
                            Submit
                        </Button>
                    </Form>
                </Card.Body>
            </Card>
        </>,
    );
}
