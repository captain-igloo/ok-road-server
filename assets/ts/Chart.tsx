import {
    BarElement,
    Chart as ChartJS,
    CategoryScale,
    LinearScale,
    PointElement,
    LineElement,
    Title,
    Tooltip,
    Legend,
} from 'chart.js';
import * as moment from 'moment';
import * as React from 'react';
import { Bar } from "react-chartjs-2";
import { useSelector } from 'react-redux';

import { RootState } from './store';

ChartJS.register(
    BarElement,
    CategoryScale,
    LinearScale,
    PointElement,
    LineElement,
    Title,
    Tooltip,
    Legend,
);

export const options = {
    maintainAspectRatio: false,
    plugins: {
      legend: {
        display: false,
      },
    },
    responsive: true,
  };

export default function Chart() {
    const features = useSelector((state: RootState) => state.okRoad.features);

    const keys = Object.keys(features);
    const values = Array(60).fill(0);
    const labels = [];

    if (keys.length > 0) {
        const diff = features[keys[keys.length - 1]].timestamp - features[keys[0]].timestamp;
        const first = features[keys[0]].timestamp;
        const a = diff / 59;
        if (a > 0) {
            keys.forEach((key) => {
                const timestamp = features[key].timestamp;
                const index = Math.floor((timestamp - first) / a);
                values[index]++;
            });

            for (let i = first; i <= features[keys[keys.length - 1]].timestamp; i += a) {
                labels.push(moment(new Date(i * 1000)).format('h:mm a'));
            }
        }
    }

    return (
        <div className="chart-container">
            <Bar options={options} data={{
                labels,
                datasets: [{
                    backgroundColor: 'rgba(255, 99, 132, 0.5)',
                    data: values,
                }],
            }} />
        </div>
    );
}