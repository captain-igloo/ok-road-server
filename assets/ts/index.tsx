import * as Sentry from '@sentry/browser';
import * as React from 'react';
import Container from 'react-bootstrap/Container';
import * as ReactDOM from 'react-dom/client';
import { Provider } from 'react-redux';

import Header from './Header';
import Important from './Important';
import { setupStore } from './store';

const element = document.getElementById('app');
if (element) {
    const dataParams = element.getAttribute('data-params');
    let configuration;
    if (dataParams) {
        configuration = JSON.parse(dataParams);
        if (configuration.sentryDsn) {
            Sentry.init({
                dsn: configuration.sentryDsn,
            });
        }
    }
    const root = ReactDOM.createRoot(element);
    root.render(
        <Provider store={setupStore()}>
            <Header user={configuration.user} />
            <Container style={{ paddingTop: '5px' }}>
                <h1>Monitor Young Drivers</h1>
                <p>
                    Install an app on their device and keep track of:
                </p>
                <ul>
                    <li>Where they are</li>
                    <li>How fast they are driving</li>
                    <li>Their speed in relation to the speed limit</li>
                </ul>
                <Important>
                    Speed limit data is currently only available in New Zealand.
                </Important>
                <h2>Getting Started</h2>
                <ol>
                    <li>
                        Click
                        {' '}
                        <a href="/register">here</a>
                        {' '}
                        to register with OK Road.  Remember your username and password because you
                        will need it in step 3.
                    </li>
                    <li>
                        Install
                        {' '}
                        <a href="https://owntracks.org">Owntracks</a>
                        {' '}
                        on the device you wish to monitor.
                    </li>
                    <li>
                        Configure Owntracks to publish to OK Road:
                        <ul>
                            <li>Set Preferences / Connection / Endpoint / Host to "192.9.176.238".</li>
                            <li>Set Preferences / Connection / Endpoint / Port to "1883".</li>
                            <li>Set Preferences / Connection / Credentials / Username to your username.</li>
                            <li>Set Preferences / Connection / Credentials / Password to your password.</li>
                        </ul>
                    </li>
                    <li>
                        Make sure that Owntracks has permission to use location services:
                        <ul>
                            <li>Set Settings / Location / Owntracks / Location access to "Allow all the time".</li>
                            <li>Set Settings / Location / Owntracks / Use precise location to true.</li>
                        </ul>
                    </li>
                </ol>
                <img alt="Screenshot" src="/img/okroad.png" />
            </Container>
        </Provider>,
    );
}
