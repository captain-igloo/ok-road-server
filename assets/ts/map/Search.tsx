import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faRightLong } from '@fortawesome/free-solid-svg-icons/faRightLong';
import { faLeftLong } from '@fortawesome/free-solid-svg-icons/faLeftLong';
import moment from 'moment';
import * as React from 'react';
import Button from 'react-bootstrap/Button';
import ButtonGroup from 'react-bootstrap/ButtonGroup';
import Card from 'react-bootstrap/Card';
import Col from 'react-bootstrap/Col';
import Form from 'react-bootstrap/Form';
import Row from 'react-bootstrap/Row';
import { useSelector } from 'react-redux';

import { RootState, useAppDispatch } from '../store';
import Devices from './Devices';
import {
    decrementDate,
    incrementDate,
    setFromDate,
    setToDate,
} from './slice';

export default function Search() {
    const fromDate = useSelector((state: RootState) => state.map.fromDate);
    const toDate = useSelector((state: RootState) => state.map.toDate);
    const dispatch = useAppDispatch();

    return (
        <Card>
            <Card.Header>Search</Card.Header>
            <Card.Body>
                <Form>
                    <Devices />
                    <Form.Group as={Row} className="mb-3">
                        <Form.Label column sm={3}>From:</Form.Label>
                        <Col sm={9}>
                            <Form.Control
                                aria-label="From"
                                onChange={(e) => {
                                    dispatch(setFromDate((new Date(e.target.value)).getTime()));
                                }}
                                type="datetime-local"
                                value={moment(new Date(fromDate)).format('Y-MM-DDTHH:mm')}
                            />
                        </Col>
                    </Form.Group>
                    <Form.Group as={Row} className="mb-3">
                        <Form.Label column sm={3}>To:</Form.Label>
                        <Col sm={9}>
                            <Form.Control
                                aria-label="To"
                                onChange={(e) => {
                                    dispatch(setToDate((new Date(e.target.value)).getTime()));
                                }}
                                type="datetime-local"
                                value={moment(new Date(toDate)).format('Y-MM-DDTHH:mm')}
                            />
                        </Col>
                    </Form.Group>
                    <Form.Group as={Row} className="mb-3">
                        <Col sm={3} />
                        <Col sm={9}>
                            <ButtonGroup>
                                <Button
                                    onClick={() => {
                                        dispatch(decrementDate());
                                    }}
                                    variant="secondary"
                                >
                                    <FontAwesomeIcon icon={faLeftLong} />
                                    <span className="display-small">
                                        &nbsp;
                                        Previous
                                    </span>
                                </Button>
                                <Button
                                    onClick={() => {
                                        dispatch(incrementDate());
                                    }}
                                    variant="secondary"
                                >
                                    <FontAwesomeIcon icon={faRightLong} />
                                    <span className="display-small">
                                        &nbsp;
                                        Next
                                    </span>
                                </Button>
                            </ButtonGroup>
                        </Col>
                    </Form.Group>
                </Form>
            </Card.Body>
        </Card>
    );
}
