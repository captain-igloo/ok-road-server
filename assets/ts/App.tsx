import * as moment from 'moment';
import * as React from 'react';
import Card from 'react-bootstrap/Card';
import Container from 'react-bootstrap/Container';
import Col from 'react-bootstrap/Col';
import Row from 'react-bootstrap/Row';
import Form from 'react-bootstrap/Form';
import { useDispatch, useSelector } from 'react-redux';

import Chart from './Chart';
import Devices from './Devices';
import Header from './Header';
import Map from './Map';
import { setFromDate, setShowRecent, setToDate } from './slice';
import { AppDispatch, RootState } from './store';

export const useAppDispatch: () => AppDispatch = useDispatch;

export default function App() {
    const bounds = useSelector((state: RootState) => state.okRoad.bounds);
    const fromDate = useSelector((state: RootState) => state.okRoad.fromDate);
    const showRecent = useSelector((state: RootState) => state.okRoad.showRecent);
    const toDate = useSelector((state: RootState) => state.okRoad.toDate);
    
    const dispatch = useAppDispatch();

    return (
        <>
            <Header />
            <Container className="main-container" fluid>
                <Row className="main-container--row">
                    <Col className="left-container" xs={3}>
                        <Card>
                            <Card.Header>Search</Card.Header>
                            <Card.Body>
                                <Form>
                                    <Devices />
                                    <Form.Group className="mb-3">
                                        <Form.Check
                                            checked={showRecent}
                                            id="show-recent"
                                            label="Last 24 hours"
                                            onChange={(e) => {
                                                dispatch(setShowRecent(e.target.checked));
                                            }}
                                            type="checkbox"
                                        />
                                    </Form.Group>
                                    <Form.Group className="mb-3">
                                        <Form.Label>From:</Form.Label>
                                        <Form.Control
                                            disabled={showRecent}
                                            onChange={(e) => {
                                                dispatch(setFromDate((new Date(e.target.value)).getTime()));
                                            }}
                                            type="datetime-local"
                                            value={moment(new Date(fromDate)).format('Y-MM-DDTHH:mm')}
                                        />
                                    </Form.Group>
                                    <Form.Group className="mb-3">
                                        <Form.Label>To:</Form.Label>
                                        <Form.Control
                                            disabled={showRecent}
                                            onChange={(e) => {
                                                dispatch(setToDate((new Date(e.target.value)).getTime()));
                                            }}
                                            type="datetime-local"
                                            value={moment(new Date(toDate)).format('Y-MM-DDTHH:mm')}
                                        />
                                    </Form.Group>
                                    <Form.Group>
                                        <Form.Check label="Fit map" type="checkbox" />
                                    </Form.Group>
                                </Form>
                            </Card.Body>
                        </Card>
                    </Col>
                    <Col className="right-container" xs={9}>
                        <Map bounds={bounds} />
                        <Chart />
                    </Col>
                </Row>
            </Container>
        </>        
    );
}
