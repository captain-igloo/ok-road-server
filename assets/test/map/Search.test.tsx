import '../setup';

import { describe, expect, test } from '@jest/globals';
import { fireEvent, render } from '@testing-library/react';
import * as React from 'react';
import { Provider } from 'react-redux';

import Search from '../../ts/map/Search';
import { setupStore } from '../../ts/store';

describe('Search component', () => {
    test('Search should render properly', () => {
        const fromDate = new Date('2025-04-16 00:00:00');
        const toDate = new Date('2025-04-17 00:00:00');
        const { container } = render(
            <Provider
                store={setupStore({
                    map: {
                        devices: [],
                        fromDate: fromDate.getTime(),
                        toDate: toDate.getTime(),
                    } as any,
                })}
            >
                <Search />
            </Provider>,
        );
        expect(container).toMatchSnapshot();
    });

    test('Change from date should set fromDate', () => {
        const store = setupStore({
            map: {
                devices: [],
                fromDate: Date.now(),
            } as any,
        });
        const { getByLabelText } = render(
            <Provider store={store}>
                <Search />
            </Provider>,
        );
        fireEvent.change(getByLabelText('From'), { target: { value: '2025-05-16T00:00' }});
        expect(store.getState().map.fromDate).toBe(1747310400000);
    });

    test('Change to date should set toDate', () => {
        const store = setupStore({
            map: {
                devices: [],
                toDate: Date.now(),
            } as any,
        });
        const { getByLabelText } = render(
            <Provider store={store}>
                <Search />
            </Provider>,
        );
        fireEvent.change(getByLabelText('To'), { target: { value: '2025-05-16T00:00' }});
        expect(store.getState().map.toDate).toBe(1747310400000);
    });
});
