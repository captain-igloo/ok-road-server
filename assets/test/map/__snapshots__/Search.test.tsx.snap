// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Search component Search should render properly 1`] = `
<div>
  <div
    class="card"
  >
    <div
      class="card-header"
    >
      Search
    </div>
    <div
      class="card-body"
    >
      <form
        class=""
      >
        <div
          class="mb-3 row"
        >
          <label
            class="form-label col-form-label col-sm-3"
          >
            Device:
          </label>
          <div
            class="col-sm-9"
          >
            <select
              aria-label="Device"
              class="form-select"
            />
          </div>
        </div>
        <div
          class="mb-3 row"
        >
          <label
            class="form-label col-form-label col-sm-3"
          >
            From:
          </label>
          <div
            class="col-sm-9"
          >
            <input
              aria-label="From"
              class="form-control"
              type="datetime-local"
              value="2025-04-16T00:00"
            />
          </div>
        </div>
        <div
          class="mb-3 row"
        >
          <label
            class="form-label col-form-label col-sm-3"
          >
            To:
          </label>
          <div
            class="col-sm-9"
          >
            <input
              aria-label="To"
              class="form-control"
              type="datetime-local"
              value="2025-04-17T00:00"
            />
          </div>
        </div>
        <div
          class="mb-3 row"
        >
          <div
            class="col-sm-3"
          />
          <div
            class="col-sm-9"
          >
            <div
              class="btn-group"
              role="group"
            >
              <button
                class="btn btn-secondary"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-left-long"
                  data-icon="left-long"
                  data-prefix="fas"
                  role="img"
                  viewBox="0 0 576 512"
                >
                  <path
                    d="M9.4 278.6c-12.5-12.5-12.5-32.8 0-45.3l128-128c9.2-9.2 22.9-11.9 34.9-6.9S192 115.1 192 128l0 64 336 0c26.5 0 48 21.5 48 48l0 32c0 26.5-21.5 48-48 48l-336 0 0 64c0 12.9-7.8 24.6-19.8 29.6s-25.7 2.2-34.9-6.9l-128-128z"
                    fill="currentColor"
                  />
                </svg>
                <span
                  class="display-small"
                >
                    Previous
                </span>
              </button>
              <button
                class="btn btn-secondary"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-right-long"
                  data-icon="right-long"
                  data-prefix="fas"
                  role="img"
                  viewBox="0 0 576 512"
                >
                  <path
                    d="M566.6 233.4c12.5 12.5 12.5 32.8 0 45.3l-128 128c-9.2 9.2-22.9 11.9-34.9 6.9S384 396.9 384 384l0-64-336 0c-26.5 0-48-21.5-48-48l0-32c0-26.5 21.5-48 48-48l336 0 0-64c0-12.9 7.8-24.6 19.8-29.6s25.7-2.2 34.9 6.9l128 128z"
                    fill="currentColor"
                  />
                </svg>
                <span
                  class="display-small"
                >
                    Next
                </span>
              </button>
            </div>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
